src/resources/2013/Florist.ts

```diff
@@ -0,0 +1,364 @@
+import {
+  floristAvailable,
+  getFloristPlants,
+  myLocation,
+  visitUrl,
+} from "kolmafia";
+import { get } from "../../property";
+
+type environment = "outdoor" | "indoor" | "underground" | "underwater";
+
+class Flower {
+  name: string;
+  id: number;
+  environment: environment;
+  modifier: string;
+  territorial: boolean;
+
+  constructor(
+    name: string,
+    id: number,
+    environment: environment,
+    modifier: string,
+    territorial = false
+  ) {
+    this.name = name;
+    this.id = id;
+    this.environment = environment;
+    this.modifier = modifier;
+    this.territorial = territorial;
+  }
+
+  isPlantedHere(location = myLocation()): boolean {
+    return getFloristPlants()[location.toString()].includes(this.name);
+  }
+
+  available(location = myLocation()): boolean {
+    return (
+      this.environment === (location.environment as environment) &&
+      !get("_floristPlantsUsed").includes(this.name) &&
+      !this.isPlantedHere(location)
+    );
+  }
+
+  dig(): boolean {
+    if (!this.isPlantedHere()) return false;
+    const flowers = getFloristPlants()[myLocation().toString()];
+    if (!flowers[2]) return false;
+    const plantNumber = getFloristPlants()[myLocation().toString()].indexOf(
+      this.name
+    );
+    visitUrl(`choice.php?option=2&whichchoice=720&pwd&plnti=${plantNumber}`);
+    return !this.isPlantedHere();
+  }
+
+  plant(): boolean {
+    if (this.isPlantedHere()) return true;
+    if (isFull()) return false;
+    visitUrl(`choice.php?whichchoice=720&whichoption=1&pwd&plant=${this.id}`);
+    return this.isPlantedHere();
+  }
+}
+
+export function have(): boolean {
+  return floristAvailable();
+}
+
+function toFlower(name: string): Flower | undefined {
+  return all().find((flower) => name === flower.name);
+}
+
+function flowersIn(location: Location): Flower[] {
+  const returnValue: Flower[] = [];
+  getFloristPlants()
+    [location.toString()].map(toFlower)
+    .forEach((flower) => {
+      if (flower) returnValue.push(flower);
+    });
+  return returnValue;
+}
+
+export function flowersAvailableFor(
+  location: Location = myLocation()
+): Flower[] {
+  return all().filter((flower) => flower.available(location));
+}
+
+export function isFull(location = myLocation()): boolean {
+  return flowersIn(location).length == 2;
+}
+
+export const all: () => Flower[] = () => [
```

Comment:
Why is this a function?

Reply:
Part if it was to prevent mutation, and part of it was because writing `FloristFriar.all()` felt natural. But yeah, there's no actual compelling reason to do so.

Reply:
Use `Object.freeze` to prevent mutation, and if you want a function thats fine but probably return an externally defined array!